subsys/sensing/sensor/phy_3d_sensor/phy_3d_sensor.c

```diff
@@ -18,65 +20,222 @@
 
 LOG_MODULE_REGISTER(phy_3d_sensor, CONFIG_SENSING_LOG_LEVEL);
 
-static int phy_3d_sensor_init(const struct device *dev,
-		const struct sensing_sensor_info *info,
-		const sensing_sensor_handle_t *reporter_handles,
-		int reporters_count)
+#define SENSING_ACCEL_Q31_SHIFT 6
+#define SENSING_GYRO_Q31_SHIFT 15
+
+static int64_t shifted_q31_to_scaled_int64(q31_t q, int8_t shift, int64_t scale)
 {
-	return 0;
+	int64_t scaled_value;
+	int64_t shifted_value;
+
+	shifted_value = (int64_t)q << shift;
+	shifted_value = llabs(shifted_value);
+
+	scaled_value =
+		FIELD_GET(GENMASK64(31 + shift, 31), shifted_value) * scale +
+		(FIELD_GET(GENMASK64(30, 0), shifted_value) * scale / BIT(31));
+
+	if (q < 0) {
+		scaled_value = -scaled_value;
+	}
+
+	return scaled_value;
 }
 
-static int phy_3d_sensor_deinit(const struct device *dev)
+static q31_t scaled_int64_to_shifted_q31(int64_t val, int64_t scale,
+		int8_t shift)
 {
-	return 0;
+	return (q31_t)((val * BIT(31 - shift) / scale));
 }
 
-static int phy_3d_sensor_read_sample(const struct device *dev,
-		void *buf, int size)
+static q31_t accel_sensor_value_to_q31(struct sensor_value *val)
 {
-	return 0;
+	int64_t micro_ms2 = sensor_value_to_micro(val);
+	int64_t micro_g = micro_ms2 * 1000000LL / SENSOR_G;
+
+	return scaled_int64_to_shifted_q31(micro_g, 1000000LL,
+			SENSING_ACCEL_Q31_SHIFT);
 }
 
-static int phy_3d_sensor_sensitivity_test(const struct device *dev,
-		int index, uint32_t sensitivity,
-		void *last_sample_buf, int last_sample_size,
-		void *current_sample_buf, int current_sample_size)
+static void accel_q31_to_sensor_value(q31_t q, struct sensor_value *val)
 {
-	return 0;
+	int64_t micro_g = shifted_q31_to_scaled_int64(q,
+			SENSING_ACCEL_Q31_SHIFT, 1000000LL);
+	int64_t micro_ms2 = micro_g * SENSOR_G / 1000000LL;
+
+	sensor_value_from_micro(val, micro_ms2);
 }
 
-static int phy_3d_sensor_set_interval(const struct device *dev, uint32_t value)
+static const struct phy_3d_sensor_custom custom_accel = {
+	.chan_all = SENSOR_CHAN_ACCEL_XYZ,
+	.shift = SENSING_ACCEL_Q31_SHIFT,
+	.q31_to_sensor_value = accel_q31_to_sensor_value,
+	.sensor_value_to_q31 = accel_sensor_value_to_q31,
+};
```

Comment:
This seems like we're still making the phy_3d_sensor an either/or. In the case of the BMI160 you'll need to support both accel and gyro here.